## Reid-Smith-Lab
Location for Reid-Smith lab projects.

This will house all bioinfomatics and data-analysis pipelines and workflows i.e. pangenome analysis, bulk RNA-seq, proteomics, genomics (WGS, Orthology etc.)

## First set-up HPC access
Let your PI know that you are requesting access (i.e. AskICT on latrobe intranet).

Introduction to using the HPC:

https://curc.readthedocs.io/en/latest/getting_started/navigating_docs.html

https://docs.erc.monash.edu/Compute/HPC/

Weird how LTU does not have its own HPC resource directory (that I know of)

Also will need to download:
1. PuTTy (HPC terminal access)
2. WinSCP (File transfer and browsing with UX interactivity to HPC and any other remote access location, i.e. you sequence something and they provide you address and key to download the raw sequencing output into your local drives)

## Download RStudio 
A lot of our scripts for local data processing are written in R, so best download R (the coding language) and Rstudio (The "environment" which the user uses to manipulate R for various analyses)
Google R and Rstudio, download them.

## Author and Contact 
Benjamin Wee Yang, PhD

B.WeeYang@latrobe.edu.au

Dugald Reid, PhD, Senior Lecturer, PI

Dugald.Reid@latrobe.edu.au
