# Pangenomics 101

What is pangenomics? Essentially, think of it as a means of representing all of the genomic variation found in a group of organisms, in contrast to a traditional reference that is (usually) assembled from a single individual.

## How do we display pangenomics infomation?

A great starting link: https://doi.org/10.1093/bib/bbae588

## First work out what kind of pangenome..
...that you are working on/want to generate.

1. Presence-abscence variation (PAV) pangenome: composed of a ‘core genome’ and an ‘accessory genome’. The core genome is the full set of genes that are present in every member of the population, while the accessory genome is composed of genes present in a subset of the population.
- focuses on gene presence and absence, does not account for gene location, allelic diversity, or intergenic sequence
2. Representative sequence pangenome: composed of carefully selected genomic sequences so that as much genomic variation from the population as possible is represented using as little sequence as possible.
  - same structure as traditional ref genome, but with aditional contigs containing supplementary genomic sequences.
3. Pangenome Graph/Graphical Pangenome: May be sequence or gene-oriented.
- Sequence-oriented pangenome graph models genomic sequence variation as well as its location relative to other genomic sequences from the population
- Gene-oriented pangenome graph models the genes found within the population and their order relative to other members of the population


## Constructing a pangenome

Two methods:
1. Homolog-based strategy: Widely used for bacteria pangenomes
2. Map-to-pan strategy

Homolog-based strategy
> 1. asdf
> 2. 
